src/ControlSystem/ControllerPND.cpp

```diff
@@ -0,0 +1,89 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ControlSystem/ControllerPND.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+namespace Controllers {
+
+template <size_t deriv_order>
+DataVector PND<deriv_order>::operator()(
+    const DataVector& timescales,
+    const std::array<DataVector, deriv_order + 1>& q_and_derivs,
+    const double q_dt, const double deriv_dt) const noexcept {
+  // helper lambda for computing the binomial coefficients
+  auto binomial = [](size_t N, size_t k) {
+    return falling_factorial(N, k) / factorial(k);
+  };
+
+  // The coefficients a0_{k} are chosen such that the system is critically
+  // damped, which are given by: a0_{k} = (N choose k) timescales^{k-N}
+  std::array<DataVector, deriv_order> coefs0;
+  const DataVector inv_timescales = 1.0 / timescales;
+  DataVector tau = inv_timescales;
+  for (size_t i = deriv_order; i-- > 0;) {
+    gsl::at(coefs0, i) = binomial(deriv_order, i) * tau;
+    // add another power of timescales
+    tau = tau * inv_timescales;
+  }
+
+  // If an averager is used for q_and_derivs, there is an induced time offset,
+  // `t_offset`, due to the time-weighted averaging. Therefore, the q_and_derivs
+  // that we have in hand are at some time `t0`. However, we desire q_and_derivs
+  // at the current time `t = t0 + t_offset` to determine the appropriate
+  // control signal. We accomplish this by Taylor expanding `q(t0 + t_offset)`.
+  // The averager allows for averaging of q and its derivatives OR to not
+  // average q while still averaging the derivatives (the derivatives are always
+  // averaged in order to reduce noise due to numerical differentiation).
+  // When they are both averaged, the time offset will be identical for q and
+  // the derivatives (`q_dt = deriv_dt`). If a 'raw' q is used, then the time
+  // offset associated with q is zero (`q_dt=0`) and the derivative time offset
+  // remains non-zero (`deriv_dt != 0)`.
+
+  // correct for time offset
+  std::array<DataVector, deriv_order> coefs;
+  for (size_t i = 0; i < deriv_order; i++) {
+    gsl::at(coefs, i) = gsl::at(coefs0, i);
+    double q_dt_fac = q_dt;
+    double deriv_dt_fac = deriv_dt;
+    double fact_denom = 1.0;
+    for (size_t j = i; j-- > 0;) {
+      // update the factorial coefficient of this term in the series expansion
+      fact_denom *= (i - j);
+      const double t_offset = (j == 0 ? q_dt_fac : deriv_dt_fac);
+      gsl::at(coefs, i) += gsl::at(coefs0, j) * t_offset / fact_denom;
+      // update the time coefficients for the next term in the series expansion
+      q_dt_fac *= q_dt;
+      deriv_dt_fac *= deriv_dt;
+    }
+  }
+
+  // compute denominator associated with correcting for time offset
+  DataVector denom{q_and_derivs[0].size(), 1.0};
+  double q_dt_fac = q_dt;
+  double deriv_dt_fac = deriv_dt;
+  double fact_denom = 1.0;
+  for (size_t i = deriv_order; i-- > 0;) {
```

Comment:
I'm unfamiliar with this style of for loop, why not `size_t i = deriv_order; i--; i > 0`? If they're functionally the same I won't make you change it, it just caught my eye

Reply:
since `i` is a size_t, but we want `i=0` to be the last value when executing the loop, it needs to be done the way it is